pippy/PipelineSchedule.py

```diff
@@ -179,19 +179,20 @@ def get_fwd_send_ops(self) -> List[dist.P2POp]:
             for fwd_output in self.fwd_outputs
         ]
 
-    def forward(self, args: List[torch.tensor]) -> torch.tensor:
-        logger.info(f"[{self.rank} FORWARD {self.stage_id}")
+    def forward(self, args: List[torch.tensor]) -> List[Any]:
         if self.is_first_stage:
             self.fwd_inputs = args
+        logger.info(
+            f"[{self.rank} FORWARD {self.stage_id} {[fwd_input.shape for fwd_input in self.fwd_inputs]}"
+        )
 
         # this is needed when we access the gradients for this in backward()
-        if not self.is_first_stage:
-            for tensor in self.fwd_inputs:
-                tensor.requires_grad = True
-                tensor.retain_grad()
+        for tensor in self.fwd_inputs:
```

Comment:
add a TODO here to make this actually depend on the requires_gradness of the actual fwd_inputs during shape inference?